### DISCUSSION TITLE
Can someone please explain how visual customization works in egui?

### DISCUSSION THREAD

**Discussion by Atlas16A** [emilk/egui#3095] [Q&A]
Starting with the Eframe Template Im working on building a gui wrapper for a cli program and ive run into customization barely working. So far i found 3 variations of customizing the look of things.

Using `let mut dark = egui::Visuals::dark();
        dark.window_fill = egui::Color32::from_rgb(42, 43, 66);
        dark.panel_fill = egui::Color32::from_rgb(42, 43, 66);
        cc.egui_ctx.set_visuals(dark);` inside the eframe::CreationContext i was able to change the global look of everything, however when i went to change panel_fill inside the SidePanel the things availble are ui.style() ui.visuals() ui.visuals_mut() and from what i read in the docs ui.visuals_mut() is what i should use however putting `ui.visuals_mut().panel_fill = egui::color32::RED` gives me nothing. If i use something like `ui.visuals_mut().override_text_color = Some(egui::color32::RED)` that works.
        
I guess what im trying to say is I dont understand what im supposed to do for assigning visuals on specific ui components and the docs just arent clear on usage.

**Comment by YgorSouza** [comment]
`ui.style_mut` and `ui.visuals_mut` affect the current `Ui` and its children. Once you are in the panel's `show` method, the current UI contains the panel's contents, not the panel itself, which is why setting the panel fill at this point doesn't work. You have to change the style before you create the panel. If the panel is created at the context level and not in an explicit UI, I think the way to do it is by doing the same thing the `settings_ui` method does: clone the style, change what you want, and then set it back in the context ([source](https://docs.rs/egui/0.22.0/src/egui/context.rs.html#1780-1786)).

```rs
let mut style = (*ctx.style()).clone();
style.visuals.panel_fill = Color32::BLUE;
ctx.set_style(style);
egui::CentralPanel::default().show(ctx, |ui| { /* ... */});
```
